File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        vector<int> prefix(heights.size());
        vector<int> suffix(heights.size());
        int ans = 0;
        stack<int> st;
        for(int i = 0; i <heights.size(); i++){
            while(!st.empty() && heights[i] <= heights[st.top()])
                st.pop();
            if(!st.empty()){
                prefix[i] = st.top() + 1;
            }
            else
                prefix[i] = 0;
            st.push(i);
        }
        
        stack<int> st1;
        for(int i = heights.size() - 1; i >= 0; i--){
            while(!st1.empty() && heights[i] <= heights[st1.top()])
                st1.pop();
            if(!st1.empty()){
                suffix[i] = st1.top() - 1;
            }
            else
                suffix[i] = heights.size() - 1;
            st1.push(i);
        }
        
        for(int i = 0; i < heights.size(); i++){
            int area = heights[i] * (suffix[i] - prefix[i] + 1);
            ans = max(ans, area);
        }
        return ans;


    }
};

File: Data Structures & Algorithms/sqrtx/submission-2.java
class Solution {
    public int mySqrt(int x) {
        if (x == 0) return 0;
        int l = 0;
        long r = x;
        long m = 0;
        int ans = 1;
        while(l<=r) {
            m = l+(r-l)/2;
            long pr = m*m;
            if(pr == x) {
                return (int)m;
            }
            else if(pr > x) {
                r = m-1;
            }
            else {
                l = (int)m+1;
                ans = (int)m;
            }
        }
        return ans;
    }
}